### PR TITLE
Resolve logic in generating the root certificate

### PR DESCRIPTION
**Pull request by krypton36**
#### Problem
When a NOC provider is passed in the root certificate is not generated. 

#### Change overview
- Add logic to check if the root certificate is present and generate it when not present. 

- Rename  variable.

- Add new variable to check if root cert is present.  

#### Testing
Ran chip-tool-darwin which passes the NOC signer.

### CODE REVIEW

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.h [hunk 1]
```diff
@@ -79,7 +79,8 @@ class CHIPOperationalCredentialsDelegate : public chip::Controller::OperationalC
     chip::NodeId mNextRequestedNodeId = 1;
     chip::FabricId mNextFabricId = 1;
     bool mNodeIdRequested = false;
-    bool mGenerateRootCert = false;
+    bool mForceRootCertRegeneration = false;
+    bool haveRootCert = false;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.h:83]: Why does `haveRootCert` need to be a member variable?

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm [hunk 2]
```diff
@@ -203,11 +203,12 @@ static BOOL isRunningTests(void)
         return CHIP_ERROR_INTERNAL;
     }
 
+    uint16_t rcacBufLen = static_cast<uint16_t>(std::min(rcac.size(), static_cast<size_t>(UINT16_MAX)));
+    PERSISTENT_KEY_OP(fabricId, kOperationalCredentialsRootCertificateStorage, key,
+        haveRootCert = mStorage->SyncGetKeyValue(key, rcac.data(), rcacBufLen) == CHIP_NO_ERROR);
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm:208]: Probably better to set `haveRootCert` ( stack variable) to false, then if `!mForceRootCertRegeneration` try doing the `SyncGetKeyValue` and set to the result of that, so we don't bother with the `SyncGetKeyValue` if we're going to ignore its result anyway?

So something like:
```
bool haveRootCert = false;
if (!mForceRootCertRegeneration) {
    // Get from storage, set haveRootCert
}

if (!haveRootCert) {
  // Generate it
}

mForceRootCertGeneration = false;
````

Comment (krypton36) [reply src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm:208]: Ah ok. Makes sense.

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.h (hunk 1 above)
Comment (krypton36) [reply src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.h]: Removed and made it local.

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm [hunk 3]
```diff
@@ -204,13 +205,14 @@ static BOOL isRunningTests(void)
...
 
     ChipDN rcac_dn;
-    if (!mGenerateRootCert) {
+    if (!mForceRootCertRegeneration) {
         uint16_t rcacBufLen = static_cast<uint16_t>(std::min(rcac.size(), static_cast<size_t>(UINT16_MAX)));
         PERSISTENT_KEY_OP(fabricId, kOperationalCredentialsRootCertificateStorage, key,
-            ReturnErrorOnFailure(mStorage->SyncGetKeyValue(key, rcac.data(), rcacBufLen)));
+            haveRootCert = mStorage->SyncGetKeyValue(key, rcac.data(), rcacBufLen) == CHIP_NO_ERROR);
         rcac.reduce_size(rcacBufLen);
         ReturnErrorOnFailure(ExtractSubjectDNFromX509Cert(rcac, rcac_dn));
```
Comment (krypton36) [inline src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm:213]: Just notice that this will fail, if there is not root cert.

Comment (krypton36) [reply src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm]: Fixed

**Review by bzbarsky-apple — APPROVED** [review]

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm [hunk 4]
```diff
@@ -204,13 +205,16 @@ static BOOL isRunningTests(void)
...
     ChipDN rcac_dn;
-    if (!mGenerateRootCert) {
+    if (!mForceRootCertRegeneration) {
         uint16_t rcacBufLen = static_cast<uint16_t>(std::min(rcac.size(), static_cast<size_t>(UINT16_MAX)));
         PERSISTENT_KEY_OP(fabricId, kOperationalCredentialsRootCertificateStorage, key,
-            ReturnErrorOnFailure(mStorage->SyncGetKeyValue(key, rcac.data(), rcacBufLen)));
-        rcac.reduce_size(rcacBufLen);
-        ReturnErrorOnFailure(ExtractSubjectDNFromX509Cert(rcac, rcac_dn));
-    } else {
+            haveRootCert = mStorage->SyncGetKeyValue(key, rcac.data(), rcacBufLen) == CHIP_NO_ERROR);
```
[the remark this answers is not in the document]
Comment (krypton36) [reply src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm]: Updated and retested. Thank you for your help on this PR.

File: src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm [hunk 5]
```diff
@@ -204,13 +205,16 @@ static BOOL isRunningTests(void)
...
+        if (haveRootCert) {
+            rcac.reduce_size(rcacBufLen);
+            ReturnErrorOnFailure(ExtractSubjectDNFromX509Cert(rcac, rcac_dn));
+        }
+    }
+    if (!haveRootCert || mForceRootCertRegeneration) {
```
[the remark this answers is not in the document]
Comment (krypton36) [reply src/darwin/Framework/CHIP/CHIPOperationalCredentialsDelegate.mm]: Updated and retested. Thank you for your help on this PR.

**Comment by bzbarsky-apple** [comment]
Fast-tracking Darwin-specific change that reverts logic to what it used to be.

**Pull request merged**